File: src/main/java/com/plugin/frege/linemarker/related/FregeFunctionBindingToAnnotationLineMarker.java
package com.plugin.frege.linemarker.related;

import com.intellij.icons.AllIcons;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.plugin.frege.psi.FregeAnnotationItem;
import com.plugin.frege.psi.impl.FregeBindingImpl;
import com.plugin.frege.psi.impl.FregeFunctionNameImpl;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.List;

public class FregeFunctionBindingToAnnotationLineMarker extends FregeRelatedItemLineMarkerAbstract {
    @Override
    protected @NotNull List<@NotNull PsiElement> getTargets(@NotNull PsiElement element) {
        PsiElement parent = element.getParent();
        if (!(parent instanceof FregeFunctionNameImpl)) {
            return List.of();
        }
        FregeBindingImpl binding = PsiTreeUtil.getParentOfType(parent, FregeBindingImpl.class);
        if (binding == null) {
            return List.of();
        }
        FregeAnnotationItem annotationItem = binding.getAnnoItem();
        return annotationItem != null ? List.of(annotationItem) : List.of();
    }

    @Override
    protected @NotNull Icon getIcon() {
        return AllIcons.Gutter.ImplementingMethod;
    }

    @Override
    protected @NotNull String getTooltipText() {
        return "Navigate to type annotation";
    }
}

File: src/main/kotlin/com/plugin/frege/linemarker/related/FregeFunctionBindingToAnnotationLineMarker.kt
package com.plugin.frege.linemarker.related

import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.plugin.frege.FregeIcons
import com.plugin.frege.psi.FregeFunctionLhs
import com.plugin.frege.psi.impl.FregeBindingImpl
import javax.swing.Icon

class FregeFunctionBindingToAnnotationLineMarker : FregeRelatedItemLineMarkerAbstract() {
    override val icon get(): Icon = FregeIcons.AnnotationMarker

    override val tooltipText get(): String = "Navigate to type annotation"

    override fun getTargets(element: PsiElement): List<PsiElement> {
        val parent = tryGetFunctionLhs(element) ?: return emptyList()
        val binding = PsiTreeUtil.getParentOfType(parent, FregeBindingImpl::class.java) ?: return emptyList()
        return if (binding.nameIdentifier === element) listOfNotNull(binding.getAnnoItem()) else emptyList()
    }

    private fun tryGetFunctionLhs(element: PsiElement): FregeFunctionLhs? {
        return generateSequence(element.parent) { it.parent }.take(3)
            .filterIsInstance<FregeFunctionLhs>().firstOrNull()
    }
}
